### PR TITLE
Enable Users To Change Sampler Using Prompts

### PR DESCRIPTION
**Pull request by blessedcoolant**
Users can now change the sampler using `-m` prompt argument. This will prevent them from having to reboot the script. Tested out all sampler changes. Could not find any issues.

**Example Usage**

`<prompt> =m k_dpm_2`

All subsequent generations will be with the same sampler until the user changes it manually or reboots the application - at which point it reverts back to the default.

### CODE REVIEW

File: scripts/dream.py [hunk 1]
```diff
@@ -508,6 +509,8 @@ def create_cmd_parser():
         action='store_true',
         help='skip subprompt weight normalization',
     )
+    parser.add_argument('-sampler', '--user_sampler', default=None, type=str, choices=['ddim', 'k_dpm_2_a', 'k_dpm_2', 'k_euler_a', 'k_euler', 'k_heun', 'k_lms', 'plms'],
```
Comment (tildebyte) [inline scripts/dream.py:512]: @blessedcoolant; Total nitpick, but will you reformat this to look like the other add_args? You could also just run 'blue' on it to fix it.

**Comment by warner-benjamin** [comment]
Good thing I checked the PRs before submitting my implementation of the same thing.

One request for your PR, could you add '-m' as an arg flag? The logged prompts use it, so would make copy/pasting old prompts easier.

Example:
`<prompt> -s30 -b1 -W512 -H704 -C7.5 -mk_lms`

File: scripts/dream.py (hunk 1 above)
Comment (blessedcoolant) [reply scripts/dream.py]: Done.

**Comment by blessedcoolant** [comment]
> Good thing I checked the PRs before submitting my implementation of the same thing.
> 
> One request for your PR, could you add '-m' as an arg flag? The logged prompts use it, so would make copy/pasting old prompts easier.
> 
> Example: `<prompt> -s30 -b1 -W512 -H704 -C7.5 -mk_lms`

Done.

File: ldm/simplet2i.py [hunk 2]
```diff
@@ -269,6 +272,11 @@ def process_image(image,seed):
 
         scope = autocast if self.precision == 'autocast' else nullcontext
 
+        if user_sampler is not None:
```
Comment (tildebyte) [inline ldm/simplet2i.py:275]: This could be refactored to

```python
if user_sampler and (user_sampler != self.sampler_name):
    etc.
```

Comment (blessedcoolant) [reply ldm/simplet2i.py:275]: I initially wrote it like that but for some reason the code was ignoring the second check, so I did a separate checker. Let me try again and see.

**Comment by tildebyte** [comment]
I feel like this arg should be brought up into `T2I.__init__()` (which for one thing should make exposing it in the web UI easier), but I can't figure out how... Either way, not **necessary** for this PR, but if you see how to do it easily, it's a "nice to have".

**Review by tildebyte — APPROVED** [review]
LGTM

**Comment by blessedcoolant** [comment]
> I feel like this arg should be brought up into `T2I.__init__()` (which for one thing should make exposing it in the web UI easier), but I can't figure out how... Either way, not **necessary** for this PR, but if you see how to do it easily, it's a "nice to have".

Thank you for reviewing the PR.

This arg is already in T2I though. But if you mean swapping it out as this PR does, then that wouldn't be possible I'm afraid. And probably not needed either. Because all the samplers use the same model, there is no need to relaunch the boot script. Any web UI should be to call and pass a prompt to T2I.prompt2png to make use of this feature.

**Comment by yunsaki** [comment]
Great stuff, changing the sampler on script start is not practical imo. I was wondering though why (in this version of stable diffusion in general) `-m` was chosen instead of `-A`. Wouldn't it be nice to maintain that backwards compatibility? Although you could argue that the syntax isn't relevant anymore since the original bot shut down. Any opinions on how to handle this?

**Comment by blessedcoolant** [comment]
> Great stuff, changing the sampler on script start is not practical imo. I was wondering though why (in this version of stable diffusion in general) `-m` was chosen instead of `-A`. Wouldn't it be nice to maintain that backwards compatibility? Although you could argue that the syntax isn't relevant anymore since the original bot shut down. Any opinions on how to handle this?

I am unsure why it was changed. But I kept it in sync with how it currently exists in this PR. I am looking at ways to optimize the prompting system too. I want all the commands to be very intuitive -- especially if the number of prompt arguments is bound to increase. But that's a discussion for another day.

**Comment by blessedcoolant** [comment]
@lstein This PR should be good to go too. It was another requested feature. No longer need to reboot script to use different samplers. Save a ton of time. Let me know if you need any changes with this one.

**Comment by lstein** [comment]
Great. I'll merge this PR first, and then fix the web UI to support both upsampling and sampler changing.

**Comment by lstein** [comment]
> Great stuff, changing the sampler on script start is not practical imo. I was wondering though why (in this version of stable diffusion in general) `-m` was chosen instead of `-A`. Wouldn't it be nice to maintain that backwards compatibility? Although you could argue that the syntax isn't relevant anymore since the original bot shut down. Any opinions on how to handle this?

I'll add -A as an alternative option. Just a brain fart.

**Comment by lstein** [comment]
Works. Only problem is that the prompt written to the file isn't updated correctly. I'm fixing.

**Comment by blessedcoolant** [comment]
> Great. I'll merge this PR first, and then fix the web UI to support both upsampling and sampler changing.

While you are doing that, maybe you could also add the option that has all the samples names in a drop down and let the users be able to pick them as they wish too.

I am currently working on the -highdetail feature and see if it's worth implementing. I am also looking at a couple of other cool functions.

But once we have a good set of features in place, I'll design a high quality full front end UI for the app using React or something.  That should make maintaining the front end really easy.

**Comment by blessedcoolant** [comment]
@lstein There's a problem with this merge. The user_sampler in simplet2i is no longer used. So it doesn't work. You need to rename the user_sampler in simplet2i to sampler_name that you set in dream.py
